internal/pgbackrest/config.go

```diff
@@ -465,6 +466,20 @@ func serverConfig(cluster *v1beta1.PostgresCluster) iniSectionSet {
 	// - https://releases.k8s.io/v1.23.0/pkg/kubelet/kubelet_pods.go#L345
 	global.Set("tls-server-address", "0.0.0.0")
 
+	// NOTE (dsessler7): As pointed out by Chris above, there is an issue in
+	// pgBackRest (#1841), where using a wildcard address to bind all addresses
+	// does not work in certain IPv6 environments. Until this is fixed, we are
+	// going to workaround the issue by allowing the user to add an annotation to
+	// enable IPv6. We will check for that annotation here and override the
+	// "tls-server-address" setting accordingly.
+	annotations := cluster.GetAnnotations()
+	if annotations != nil {
+		if ipVersion, exists := annotations[naming.PGBackRestIPVersion]; exists &&
+			strings.ToLower(ipVersion) == "ipv6" {
+			global.Set("tls-server-address", "::")
+		}
+	}
```

Comment:
I was definitely aware of the fact that in Go the zero value will be returned when attempting to retrieve a value that doesn't exist from a map, so I'm honestly not sure why I didn't use that functionality when retrieving the specific annotation and comparing its value to "ipv6". Guess it just slipped my mind.

However, in regards to retrieving the annotations themselves, I think I noticed the usage of the `GetAnnotations()` and `GetAnnotationsOrNil()` functions throughout the codebase and assumed, perhaps incorrectly, that one should be used when retrieving annotations and that if no annotations had been set that the return value would be `nil` (and attempting to retrieve a value from `nil` would throw an error). Hence the initial check for its existence.

I think I'm also slightly confused about the structure of the PostgresCluster struct. I would not have thought to access Annotations directly from the cluster since they reside inside the cluster's metadata. Also, when I look at the PostgresCluster struct I see ObjectMeta and TypeMeta, which further confuse me, but I also haven't taken the time to dig in and figure out the difference and how they each work.

Reply:
> I think I'm also slightly confused about the structure of the PostgresCluster struct. I would not have thought to access Annotations directly from the cluster since they reside inside the cluster's metadata.

These are [embedded fields](https://go.dev/ref/spec#Struct_types) because they don't have a field name in front of the field type. The methods you're calling are actually being called on those.

https://github.com/CrunchyData/postgres-operator/blob/607c1b1ae80427cfdb93e96a54293446990c3eb2/pkg/apis/postgres-operator.crunchydata.com/v1beta1/postgrescluster_types.go#L596-L597

> Also, when I look at the PostgresCluster struct I see ObjectMeta and TypeMeta, which further confuse me, but I also haven't taken the time to dig in and figure out the difference and how they each work.

ObjectMeta is name, namespace, etc. TypeMeta is the GVK. The `json:",inline"` tag tells the JSON encoder/decoder to put... that field's fields... up one level.

- https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ObjectMeta
- https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#TypeMeta